console.go

```diff
@@ -0,0 +1,165 @@
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/miquella/ask"
+	"github.com/miquella/vaulted/lib"
+	"github.com/pkg/browser"
+)
+
+var (
+	ErrInvalidTemporaryCredentials = errors.New("Invalid temporary session credentials present, permanent credentials are required.\nPermanent credentials can be used be specifying a vault name")
+	ErrInvalidDuration             = errors.New("Minimum console duration must be at least 15 min")
+)
+
+const (
+	CONSOLE_URL = "https://console.aws.amazon.com/console/home"
+	SIGNIN_URL  = "https://signin.aws.amazon.com/federation"
+
+	MinConsoleDuration = 15 * time.Minute
+	MaxConsoleDuration = 12 * time.Hour
+)
+
+type Console struct {
+	VaultName string
+	Role      string
+}
+
+func (c *Console) Run(steward Steward) error {
+	awsCreds, duration, err := c.getCreds(steward)
+	if err != nil {
+		return err
+	}
+
+	signinToken, err := awsCreds.GetSigninToken(duration)
+	if err != nil {
+		return err
+	}
+
+	// console signin
+	signinUrl, _ := url.Parse(SIGNIN_URL)
+	loginQuery := make(url.Values)
+	loginQuery.Set("Action", "login")
+	loginQuery.Set("SigninToken", signinToken)
+	loginQuery.Set("Destination", CONSOLE_URL)
+
+	signinUrl.RawQuery = loginQuery.Encode()
+	err = browser.OpenURL(signinUrl.String())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Console) getCreds(steward Steward) (*vaulted.AWSCredentials, *time.Duration, error) {
+	var err error
+	var awsCreds *vaulted.AWSCredentials
+	var duration *time.Duration
+
+	// default session
+	session := DefaultSession()
+
+	if c.VaultName != "" {
+		// get specific vault
+		_, v, err := steward.OpenVault(c.VaultName, nil)
+		if err != nil {
+			return nil, nil, err
+		}
+		awsCreds = &v.AWSKey.AWSCredentials
+
+		if c.Role != "" || (v.AWSKey != nil && v.AWSKey.Role != "") {
+			var role string
+			if role = c.Role; role == "" {
+				role = v.AWSKey.Role
+			}
+
+			if v.AWSKey.MFA != "" {
+				tokenCode, err := c.getTokenCode()
+				if err != nil {
+					return nil, nil, err
+				}
+				awsCreds, err = awsCreds.AssumeRoleWithMFA(v.AWSKey.MFA, tokenCode, role, MinConsoleDuration)
+			} else {
+				awsCreds, err = awsCreds.AssumeRole(role, MinConsoleDuration)
+				if err != nil {
+					return nil, nil, err
+				}
+			}
+
+			duration, err = getValidDuration(v.Duration)
+			if err != nil {
+				return nil, nil, err
+			}
+		} else {
+			fedDuration, err := getValidDuration(v.Duration)
+			if err != nil {
+				return nil, nil, err
+			}
+			awsCreds, err = awsCreds.GetFederationToken(*fedDuration)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
+	} else {
+		tempCreds, err := session.AWSCreds.GetLoadedOrDefaultCreds()
+		if err != nil || tempCreds.Token != "" {
+			return nil, nil, ErrInvalidTemporaryCredentials
+		}
+
+		fedDuration, err := getValidDuration(session.Expiration.Sub(time.Now()))
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if c.Role != "" {
+			awsCreds, err = session.AWSCreds.AssumeRole(c.Role, *fedDuration)
```

Comment:
Perhaps we can try to merge some of these checks / assumes?  We seem to be repeating certain actions between vault and no-vault, I think we can wrap them a bit cleaner.